### PR TITLE
Handle begin & end block channel events

### PR DESCRIPTION
**Pull request by hu55a1n1**
Closes: #1793 

## Description
This PR is mostly based on the code that was removed in PR #1172. 
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [x] Reviewed `Files changed` in the GitHub PR explorer.
- [x] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by hu55a1n1** [comment]
Tested `hermes listen` with the following patch (https://github.com/agoric-labs/cosmos-sdk/tree/mfig-endblock-ibc-event) and it seems to work. 
```diff
diff --git a/x/staking/module.go b/x/staking/module.go
index b5e778755..624b33a49 100644
--- a/x/staking/module.go
+++ b/x/staking/module.go
@@ -23,6 +23,7 @@ import (
        "github.com/cosmos/cosmos-sdk/x/staking/keeper"
        "github.com/cosmos/cosmos-sdk/x/staking/simulation"
        "github.com/cosmos/cosmos-sdk/x/staking/types"
+       channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
 )
 
 var (
@@ -171,12 +172,21 @@ func (am AppModule) BeginBlock(ctx sdk.Context, _ abci.RequestBeginBlock) {
 // EndBlock returns the end blocker for the staking module. It returns no validator
 // updates.
 func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
-       // FIGME: Raise an event in end block.
-       AttributeValueCategory := "ibc_channel"
+       ctx.EventManager().EmitEvents(sdk.Events{
+               sdk.NewEvent(
+                       channeltypes.EventTypeChannelOpenInit,
+                       sdk.NewAttribute(channeltypes.AttributeKeyPortID, "transfer"),
+                       sdk.NewAttribute(channeltypes.AttributeKeyChannelID, "channel-0"),
+                       sdk.NewAttribute(channeltypes.AttributeCounterpartyPortID, "transfer"),
+                       sdk.NewAttribute(channeltypes.AttributeCounterpartyChannelID, "channel-0"),
+                       sdk.NewAttribute(channeltypes.AttributeKeyConnectionID, "connection-0"),
+               ),
+       })
+
        ctx.EventManager().EmitEvents(sdk.Events{
                sdk.NewEvent(
                        sdk.EventTypeMessage,
-                       sdk.NewAttribute(sdk.AttributeKeyModule /* channeltypes. */, AttributeValueCategory),
+                       sdk.NewAttribute(sdk.AttributeKeyModule, channeltypes.AttributeValueCategory),
                ),
        })
        return EndBlocker(ctx, am.keeper)
```

File: relayer/src/event/rpc.rs [hunk 1]
```diff
@@ -82,19 +75,78 @@ pub fn get_all_events(
...
+        &mut events,
+        extract_events(height, block_events, "channel_open_ack", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::OpenConfirm>(
+        &mut events,
+        extract_events(height, block_events, "channel_open_confirm", "channel_id"),
+        height,
+    );
+    append_events::<ChannelEvents::SendPacket>(
```
Comment (hu55a1n1) [inline relayer/src/event/rpc.rs:136]: Note: only `send_packet` events are handled. Should we handle others?

Comment (ancazamfir) [reply relayer/src/event/rpc.rs:136]: We could but I don't see a requirement. Even for the channel events other than open_init and close_init not sure it makes sense to have. When generated in the begin/end block the implication is that the app has somehow access to the other chain, able to get proofs, etc.

**Review by ancazamfir — APPROVED** [review]
great stuff! LGTM!

File: relayer/src/event/rpc.rs [hunk 2]
```diff
@@ -11,6 +11,106 @@ use ibc::events::{IbcEvent, RawObject};
...
+///
+/// Events originate from the following ABCI methods ->
+/// 1. `DeliverTx` - these events are generated during the execution of transaction messages.
+/// 2. `BeginBlock`
+/// 3. `EndBlock`
+///
+/// Events originating from `DeliverTx` are currently extracted via the `RpcEvent::data` field as
+/// the `EventData::Tx` variant.
+/// Here's an example of what these events look like (i.e. `TxInfo::TxResult::events`) -
+/// ```ron
```
Comment (adizere) [inline relayer/src/event/rpc.rs:24]: nit: json instead of ron?

**Review by adizere — APPROVED** [review]
Very clear explanation, thank you Shoaib!

I think the PR is ready, but let's let @michaelfig test it out against Agoric devnet and give the final approval.

**Comment by michaelfig** [comment]
> I think the PR is ready, but let's let @michaelfig test it out against Agoric devnet and give the final approval.

Thanks, everybody, for your work on this!

This is high up on my list of priorities.  @dckc and I are doing IBC testing, so we'll be able to try a programmatic outbound channel init from an Agoric (dynamic IBC) smart contract in the next day or three.

Will keep you posted!

**Comment by dckc** [comment]
good news / bad news:

We were able to initiate a channel using JavaScript:

```
hermes_1  | 2022-02-02T15:55:16.352132Z DEBUG ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: starts processing Some(
hermes_1  |     OpenInitChannel(
hermes_1  |         OpenInit(
hermes_1  |             Attributes {
hermes_1  |                 height: Height {
hermes_1  |                     revision: 0,
hermes_1  |                     height: 11213,
hermes_1  |                 },
hermes_1  |                 port_id: PortId(
hermes_1  |                     "port-4",
hermes_1  |                 ),
hermes_1  |                 channel_id: Some(
hermes_1  |                     ChannelId(
hermes_1  |                         "channel-2",
hermes_1  |                     ),
hermes_1  |                 ),
hermes_1  |                 connection_id: ConnectionId(
hermes_1  |                     "connection-1",
hermes_1  |                 ),
hermes_1  |                 counterparty_port_id: PortId(
hermes_1  |                     "transfer",
hermes_1  |                 ),
hermes_1  |                 counterparty_channel_id: None,
hermes_1  |             },
hermes_1  |         ),
hermes_1  |     ),
hermes_1  | )
```

but then

```
2022-02-02T15:55:21.496369Z DEBUG ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: resolving channel version by retrieving destination chain app version chain_id=cosmoshub-testnet port_id=transfer ctx=ChanOpenTry


2022-02-02T15:55:57.279753Z  WARN ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: cannot resolve channel version for unknown port id 'port-4'

hermes_1  | 2022-02-02T15:55:55.768779Z TRACE ThreadId(18) fetching header height=0-204
hermes_1  | 2022-02-02T15:55:57.279380Z DEBUG ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: [agoric -> cosmoshub-testnet:07-tendermint-141] MsgUpdateAnyClient from trusted height 0-203 to target height 0-11221
hermes_1  | 2022-02-02T15:55:57.279575Z DEBUG ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: resolving channel version by retrieving destination chain app version chain_id=cosmoshub-testnet port_id=transfer ctx=ChanOpenTry
hermes_1  | 2022-02-02T15:55:57.279753Z  WARN ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: cannot resolve channel version for unknown port id 'port-4'
hermes_1  | 2022-02-02T15:55:57.279872Z ERROR ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: Failed ChanInit with error: failed during a query for the app version to chain id cosmoshub-testnet: failed to fetch application version: could not resolve channel version because the port is invalid: port-4
hermes_1  | 2022-02-02T15:55:57.279946Z ERROR ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: aborting after encountering fatal error: Worker failed after 7 retries
hermes_1  | 2022-02-02T15:55:57.279978Z  INFO ThreadId(35) channel{channel=channel::channel-2/port-4:agoric -> cosmoshub-testnet}: terminated
```

**Comment by michaelfig** [comment]
In short, this particular PR looks good and does what it says it should.

Go ahead and land!

File: relayer/src/event/rpc.rs (hunk 2 above)
Comment (hu55a1n1) [reply relayer/src/event/rpc.rs:24]: Sorry for not addressing this - the snippet is just debug output, so it's closer to `ron` than it is to `json`.

Comment (romac) [reply relayer/src/event/rpc.rs:24]: No worries, it's fine as is imho

**Pull request merged**